diff --exclude '*.o' --exclude '*.mod' --exclude 'lib*.a' --exclude Makefile --exclude FoX_config -C 3 -r FoX-2.1.1/sax/m_sax_parser.f90 FoX-2.1.1-patched/sax/m_sax_parser.f90
*** FoX-2.1.1/sax/m_sax_parser.f90	2007-05-25 07:30:37.000000000 -0400
--- FoX-2.1.1-patched/sax/m_sax_parser.f90	2007-05-28 11:38:00.000000000 -0400
***************
*** 59,64 ****
--- 59,65 ----
      call init_element_list(fx%element_list)
  
      allocate(fx%wf_stack(1))
+ fx%wf_stack(1) = 0
      call init_entity_list(fx%forbidden_ge_list)
      call init_entity_list(fx%forbidden_pe_list)
  
diff --exclude '*.o' --exclude '*.mod' --exclude 'lib*.a' --exclude Makefile --exclude FoX_config -C 3 -r FoX-2.1.1/sax/m_sax_reader.f90 FoX-2.1.1-patched/sax/m_sax_reader.f90
*** FoX-2.1.1/sax/m_sax_reader.f90	2007-05-25 07:30:37.000000000 -0400
--- FoX-2.1.1-patched/sax/m_sax_reader.f90	2007-05-28 12:34:24.000000000 -0400
***************
*** 168,174 ****
      fb%eof = .false.
      fb%line = 1
      fb%col = 0
!     fb%filename => vs_str_alloc(file)
      fb%pos = 1
      fb%nchars = 0
      allocate(fb%buffer_stack(0))
--- 168,179 ----
      fb%eof = .false.
      fb%line = 1
      fb%col = 0
!     if (present(file)) then
!       fb%filename => vs_str_alloc(file)
!     else
!       fb%filename => vs_str_alloc("")
!     endif
! 
      fb%pos = 1
      fb%nchars = 0
      allocate(fb%buffer_stack(0))
