Index: fft_lib/OBJECTDEFS
===================================================================
RCS file: /cvsroot/cp2k/cp2k/src/fft_lib/OBJECTDEFS,v
retrieving revision 1.2
diff -b -c -3 -r1.2 OBJECTDEFS
*** fft_lib/OBJECTDEFS	6 Jan 2011 08:23:15 -0000	1.2
--- fft_lib/OBJECTDEFS	3 May 2011 15:07:08 -0000
***************
*** 1,4 ****
--- 1,5 ----
  LIB2_OBJECTS = ctrig.o\
+                fft_kinds_pre.o\
                 fft_kinds.o\
                 fft_lib.o\
                 fft_plan.o\
***************
*** 15,17 ****
--- 16,20 ----
                 fftw3_lib.o\
                 mltfftsg_tools.o\
                 mltfftsg.o
+ 
+ fft_kinds.o : fft_kinds_pre.o
Index: fft_lib/fft_kinds.F
===================================================================
RCS file: /cvsroot/cp2k/cp2k/src/fft_lib/fft_kinds.F,v
retrieving revision 1.2
diff -b -c -3 -r1.2 fft_kinds.F
*** fft_lib/fft_kinds.F	31 Jan 2011 07:52:12 -0000	1.2
--- fft_lib/fft_kinds.F	3 May 2011 15:07:08 -0000
***************
*** 7,15 ****
  !> \brief Defines the basic variable types
  !> \author Teodoro Laino - 01.2011
  ! *****************************************************************************
! #if defined ( __FFTMKL )
! #include 'mkl_dfti.f90'
! #endif
   
  MODULE fft_kinds
  
--- 7,16 ----
  !> \brief Defines the basic variable types
  !> \author Teodoro Laino - 01.2011
  ! *****************************************************************************
! ! moved to fft_kinds_pre.F by NB (needs to be compiled first so that MKL_DFTI module exists)
! ! #if defined ( __FFTMKL )
! ! #include 'mkl_dfti.f90'
! ! #endif
   
  MODULE fft_kinds
  
diff -N -b -c -r fft_lib/fft_kinds_pre.F fft_lib/fft_kinds_pre.F
*** fft_lib/fft_kinds_pre.F	1969-12-31 14:00:00.000000000 -1000
--- fft_lib/fft_kinds_pre.F	2011-05-03 05:42:47.000000000 -1000
***************
*** 0 ****
--- 1,3 ----
+ #if defined ( __FFTMKL )
+ #include "mkl_dfti.f90"
+ #endif

