Index: input_constants.F
===================================================================
RCS file: /cvsroot/cp2k/cp2k/src/input_constants.F,v
retrieving revision 1.292
diff -u -r1.292 input_constants.F
--- input_constants.F	17 Jan 2011 16:50:29 -0000	1.292
+++ input_constants.F	31 Jan 2011 22:22:49 -0000
@@ -327,7 +327,8 @@
                                               do_coord_crd=4,&
                                               do_coord_cif=5,&
                                               do_coord_xtl=6,&
-                                              do_coord_cp2k=7
+                                              do_coord_cp2k=7,&
+                                              do_coord_exyz=8
 
   INTEGER, PARAMETER, PUBLIC               :: do_ff_undef=0,&
                                               do_ff_quartic=1,&
Index: input_cp2k_subsys.F
===================================================================
RCS file: /cvsroot/cp2k/cp2k/src/input_cp2k_subsys.F,v
retrieving revision 1.138
diff -u -r1.138 input_cp2k_subsys.F
--- input_cp2k_subsys.F	17 Jan 2011 16:50:29 -0000	1.138
+++ input_cp2k_subsys.F	31 Jan 2011 22:22:49 -0000
@@ -378,6 +378,11 @@
             default_l_val=.FALSE.,lone_keyword_l_val=.TRUE.,error=error)
        CALL section_add_keyword(print_key,keyword,error=error)
        CALL keyword_release(keyword,error=error)
+       CALL keyword_create(keyword,"exyz_info",&
+            description="Prints information when parsing EXYZ files.",&
+            default_l_val=.FALSE.,lone_keyword_l_val=.TRUE.,error=error)
+       CALL section_add_keyword(print_key,keyword,error=error)
+       CALL keyword_release(keyword,error=error)
        CALL keyword_create(keyword,"psf_info",&
             description="Prints information when parsing PSF files.",&
             default_l_val=.FALSE.,lone_keyword_l_val=.TRUE.,error=error)
@@ -1570,10 +1575,10 @@
        CALL keyword_create(keyword, name="COORD_FILE_FORMAT",&
             variants=s2a("COORDINATE"),&
             description="Set up the way in which coordinates will be read.",&
-            usage="COORD_FILE_FORMAT (OFF|PDB|XYZ|G96|CRD|CIF|XTL|CP2K)", &
-            enum_c_vals=s2a( "OFF","PDB","XYZ","G96","CRD","CIF","XTL","CP2K"),&
+            usage="COORD_FILE_FORMAT (OFF|PDB|XYZ|G96|CRD|CIF|XTL|CP2K|EXYZ)", &
+            enum_c_vals=s2a( "OFF","PDB","XYZ","G96","CRD","CIF","XTL","CP2K","EXYZ"),&
             enum_i_vals=(/do_coord_off, do_coord_pdb, do_coord_xyz, do_coord_g96, do_coord_crd,&
-            do_coord_cif,do_coord_xtl,do_coord_cp2k/),&
+            do_coord_cif,do_coord_xtl,do_coord_cp2k,do_coord_exyz/),&
             enum_desc=s2a(&
             "Coordinates read in the &COORD section of the input file",&
             "Coordinates provided through a PDB file format",&
@@ -1582,7 +1587,8 @@
             "Coordinates provided through an AMBER file format",&
             "Coordinates provided through a CIF (Crystallographic Information File) file format",&
             "Coordinates provided through a XTL (MSI native) file format",&
-            "Read the coordinates in CP2K &COORD section format from an external file."),&
+            "Read the coordinates in CP2K &COORD section format from an external file,",&
+            "Coordinates provided through an EXYZ (QUIP native) file format."),&
             default_i_val=do_coord_off,error=error)
        CALL section_add_keyword(section,keyword,error=error)
        CALL keyword_release(keyword,error=error)
Index: topology.F
===================================================================
RCS file: /cvsroot/cp2k/cp2k/src/topology.F,v
retrieving revision 1.147
diff -u -r1.147 topology.F
--- topology.F	17 Jan 2011 16:50:29 -0000	1.147
+++ topology.F	31 Jan 2011 22:22:49 -0000
@@ -27,7 +27,7 @@
        do_conn_amb7, do_conn_g87, do_conn_g96, do_conn_generate, &
        do_conn_mol_set, do_conn_off, do_conn_psf, do_conn_psf_u, &
        do_conn_user, do_coord_cif, do_coord_cp2k, do_coord_crd, do_coord_g96, &
-       do_coord_off, do_coord_pdb, do_coord_xtl, do_coord_xyz
+       do_coord_off, do_coord_pdb, do_coord_xtl, do_coord_xyz, do_coord_exyz
   USE input_section_types,             ONLY: section_vals_get,&
                                              section_vals_get_subs_vals,&
                                              section_vals_type,&
@@ -87,6 +87,7 @@
                                              topology_set_atm_mass
   USE topology_xtl,                    ONLY: read_coordinate_xtl
   USE topology_xyz,                    ONLY: read_coordinate_xyz
+  USE topology_exyz,                   ONLY: read_coordinate_exyz
 #include "cp_common_uses.h"
 
   IMPLICIT NONE
@@ -554,6 +555,8 @@
        CALL read_coordinate_pdb (topology,para_env,subsys_section,error)
     CASE (do_coord_xyz)
        CALL read_coordinate_xyz (topology,para_env,subsys_section,error)
+    CASE (do_coord_exyz)
+       CALL read_coordinate_exyz (topology,para_env,subsys_section,error)
     CASE (do_coord_cif)
        CALL read_coordinate_cif (topology,para_env,subsys_section,error)
     CASE (do_coord_xtl)
Index: OBJECTDEFS
===================================================================
RCS file: /cvsroot/cp2k/cp2k/src/OBJECTDEFS,v
retrieving revision 1.537
diff -u -r1.537 OBJECTDEFS
--- OBJECTDEFS	17 Jan 2011 16:50:29 -0000	1.537
+++ OBJECTDEFS	31 Jan 2011 22:22:49 -0000
@@ -657,6 +657,7 @@
  topology_util.o\
  topology_xtl.o\
  topology_xyz.o\
+ topology_exyz.o\
  util.o\
  velocity_verlet_control.o\
  vibrational_analysis.o\
